File: cli/lib/review-workflow.mjs
import { inspectHtml } from "./design-quality-inspector.mjs";
import {
  canonicalReviewStages,
  reviewSourceDigest,
  reviewWorkflowDigest,
  validateReviewWorkflow,
} from "./review-workflow-contract.mjs";
import { buildStartPayload } from "./start-operation.mjs";

function requiredText(value, field) {
  if (typeof value !== "string" || value.trim() === "") {
    throw new Error(`${field} must be a non-empty string`);
  }
  return value.trim();
}

function optionalText(value, field) {
  if (value === undefined || value === null || value === "") return "";
  return requiredText(value, field);
}

function textList(value, field, fallback = []) {
  if (value === undefined || value === null) return [...fallback];
  if (!Array.isArray(value)) throw new Error(`${field} must be an array`);
  const items = value.map((item, index) => requiredText(item, `${field}[${index}]`));
  return [...new Set(items)];
}

export function buildReviewWorkflow(source, options) {
  if (typeof source !== "string" || source.trim() === "") {
    throw new Error("source must be a non-empty string");
  }
  const html = source;
  if (!options || typeof options !== "object" || Array.isArray(options)) {
    throw new Error("review options must be an object");
  }

  const brief = requiredText(options.brief, "brief");
  const sourceRef = requiredText(options.sourceRef, "sourceRef");
  const locale = optionalText(options.locale, "locale") || "en";
  const viewports = textList(options.viewports, "viewports", ["mobile", "desktop"]);
  if (viewports.length === 0) throw new Error("viewports must be a non-empty array");

  const plan = buildStartPayload({
    brief,
    sourceRoot: requiredText(options.sourceRoot, "sourceRoot"),
    prefix: requiredText(options.prefix, "prefix"),
    routeId: "design-review",
    context: {
      siteName: optionalText(options.siteName, "siteName"),
      repoUrl: optionalText(options.repoUrl, "repoUrl"),
      localPath: optionalText(options.localPath, "localPath"),
      url: optionalText(options.url, "url"),
      screenshots: textList(options.screenshots, "screenshots"),
      locale,
      viewports,
    },
  });
  const report = inspectHtml(html, {
    sourceRef,
    brief,
    name: optionalText(options.name, "name") || undefined,
    locale,
    viewports,
    generatedAt: optionalText(options.generatedAt, "generatedAt") || undefined,
    reviewPack: optionalText(options.reviewPack, "reviewPack") || undefined,
  });

  const workflow = {
    kind: "design-ai-review-workflow",
    schemaVersion: 1,
    status: "static-review-complete",
    source: {
      reference: sourceRef,
      sha256: reviewSourceDigest(html),
      bytes: Buffer.byteLength(html, "utf8"),
    },
    plan,
    report,
    linkage: {
      status: "pass",
      briefMatch: true,
      localeMatch: true,
      viewportMatch: true,
      sourceReferenceMatch: true,
      planSha256: reviewWorkflowDigest(plan),
      designContractSha256: reviewWorkflowDigest(plan.designContract),
      reportSha256: reviewWorkflowDigest(report),
    },
    stages: canonicalReviewStages(),
    nextAction: {
      id: "human-review-required",
      status: "pending",
      summary: report.summary.nextAction,
      approvalRequiredBefore: [...report.approval.requiredBefore],
    },
    boundary: {
      mode: "read-only",
      localWrites: false,
      targetRepoMutation: false,
      externalWrites: false,
    },
  };

  return validateReviewWorkflow(workflow, { source: html });
}

File: tools/audit/smoke_domains/review_workflow.py
"""Canonical P6 review-workflow smoke assertion."""
from __future__ import annotations

import hashlib
import json
from pathlib import Path

from .assertion_helpers import assert_no_ansi, assert_smoke_json_keys
from .review_quality import review_workflow_digest


def assert_review_workflow_json(
    raw: str,
    source_path: Path,
    *,
    context: str,
    cmd: list[str],
) -> None:
    assert_no_ansi(raw, cmd)
    try:
        workflow = json.loads(raw)
    except json.JSONDecodeError as error:
        raise SystemExit(f"failed to parse review workflow JSON after {context}") from error

    assert_smoke_json_keys(
        workflow,
        [
            "kind", "schemaVersion", "status", "source", "plan", "report",
            "linkage", "stages", "nextAction", "boundary",
        ],
        label="top-level",
        context=context,
        command_label="review workflow JSON",
    )
    if (
        workflow.get("kind") != "design-ai-review-workflow"
        or workflow.get("schemaVersion") != 1
        or workflow.get("status") != "static-review-complete"
    ):
        raise SystemExit(f"review workflow JSON after {context} changed its contract identity")

    source_bytes = source_path.read_bytes()
    source = workflow.get("source")
    expected_source = {
        "reference": str(source_path.resolve()),
        "sha256": hashlib.sha256(source_bytes).hexdigest(),
        "bytes": len(source_bytes),
    }
    if source != expected_source:
        raise SystemExit(f"review workflow JSON after {context} lost exact source identity")

    plan = workflow.get("plan")
    report = workflow.get("report")
    if not (
        isinstance(plan, dict)
        and plan.get("kind") == "design-ai-start"
        and plan.get("schemaVersion") == 1
        and isinstance(plan.get("route"), dict)
        and plan["route"].get("id") == "design-review"
        and isinstance(report, dict)
        and report.get("kind") == "design-ai-quality-report"
        and report.get("schemaVersion") == 1
        and isinstance(report.get("context"), dict)
        and report["context"].get("routeId") == "design-engineering-review"
    ):
        raise SystemExit(f"review workflow JSON after {context} changed its nested artifact identities")

    summary = report.get("summary")
    if not (
        isinstance(summary, dict)
        and summary.get("status") == "fail"
        and summary.get("confirmedFindings") == 1
        and summary.get("unverifiedFindings") == 1
    ):
        raise SystemExit(f"review workflow JSON after {context} lost confirmed and unverified evidence")

    linkage = workflow.get("linkage")
    expected_linkage = {
        "status": "pass",
        "briefMatch": True,
        "localeMatch": True,
        "viewportMatch": True,
        "sourceReferenceMatch": True,
        "planSha256": review_workflow_digest(plan),
        "designContractSha256": review_workflow_digest(plan.get("designContract")),
        "reportSha256": review_workflow_digest(report),
    }
    if linkage != expected_linkage:
        raise SystemExit(f"review workflow JSON after {context} changed artifact linkage evidence")

    expected_stages = [
        {"id": "plan", "status": "complete", "artifactKind": "design-ai-start"},
        {"id": "static-review", "status": "complete", "artifactKind": "design-ai-quality-report"},
        {"id": "browser-verification", "status": "not-run", "artifactKind": None},
        {"id": "implementation-handoff", "status": "not-started", "artifactKind": None},
    ]
    if workflow.get("stages") != expected_stages:
        raise SystemExit(f"review workflow JSON after {context} changed the canonical stage sequence")

    approval = report.get("approval")
    required_before = approval.get("requiredBefore") if isinstance(approval, dict) else None
    next_action = workflow.get("nextAction")
    if next_action != {
        "id": "human-review-required",
        "status": "pending",
        "summary": summary.get("nextAction"),
        "approvalRequiredBefore": required_before,
    }:
        raise SystemExit(f"review workflow JSON after {context} changed the human review gate")

    if workflow.get("boundary") != {
        "mode": "read-only",
        "localWrites": False,
        "targetRepoMutation": False,
        "externalWrites": False,
    }:
        raise SystemExit(f"review workflow JSON after {context} changed its read-only boundary")
